File: Leetcode/42_Trapping Rain Water.c
int trap(int* height, int heightSize) {
    if (heightSize <= 2) return 0;

    int n = heightSize;
    int lmax = 0, rmax = 0;

    int max_l[n];
    int max_r[n];
    for (int i = 0; i < n; i++) {
        max_l[i] = lmax;
        if (height[i] > lmax)
            lmax = height[i];

        max_r[n - 1 - i] = rmax;
        if (height[n - 1 - i] > rmax)
            rmax = height[n - 1 - i];
    }

    int total = 0;

    for (int i = 0; i < n; i++) {
        int water_level = (max_l[i] < max_r[i]) ? max_l[i] : max_r[i];
        int above_water = water_level - height[i];

        if (above_water > 0)
            total += above_water;
    }

    return total;
}

File: Leetcode/42_Trapping Rain Water.py
class Solution:
    def trap(self, height):
        n = len(height)
        

        lmax = 0
        rmax = 0

        max_l = [0] * n
        max_r = [0] * n

      
        for i in range(n):
            j=-i-1
            max_l[i]=lmax
            max_r[j]=rmax

            lmax=max(lmax,height[i])
            
            rmax=max(rmax,height[j])

        total = 0
        for i in range(n):
            water_level = min(max_l[i], max_r[i])
            above_water = water_level - height[i]

            if above_water > 0:
                total += above_water

        return total
